File: shared/src/main/java/net/katsstuff/old/chitchat/ChitChat.java
/**
 * This file is part of ChitChat, licensed under the MIT License (MIT).
 *
 * Copyright (c) 2016 Katrix
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
 * associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute,
 * sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
 * NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
 * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package net.katsstuff.old.chitchat;

import java.io.IOException;
import java.nio.file.Path;

import org.slf4j.Logger;
import org.spongepowered.api.Sponge;
import org.spongepowered.api.config.ConfigDir;
import org.spongepowered.api.data.DataManager;
import org.spongepowered.api.event.Listener;
import org.spongepowered.api.event.game.GameReloadEvent;
import org.spongepowered.api.event.game.state.GameConstructionEvent;
import org.spongepowered.api.event.game.state.GameInitializationEvent;
import org.spongepowered.api.plugin.Plugin;
import org.spongepowered.api.plugin.PluginContainer;

import com.google.common.reflect.TypeToken;
import com.google.inject.Inject;

import net.katsstuff.old.chitchat.chat.ChannelChitChat;
import net.katsstuff.old.chitchat.chat.ChannelChitChatSerializer;
import net.katsstuff.old.chitchat.chat.ChannelRootSerializer;
import net.katsstuff.old.chitchat.chat.ChatListener;
import net.katsstuff.old.chitchat.chat.data.ImmutableUserChitChatData;
import net.katsstuff.old.chitchat.chat.data.UserChitChatData;
import net.katsstuff.old.chitchat.chat.data.UserChitChatManipulatorBuilder;
import net.katsstuff.old.chitchat.command.CmdAnnounce;
import net.katsstuff.old.chitchat.command.CmdChannel;
import net.katsstuff.old.chitchat.command.CmdChitChat;
import net.katsstuff.old.chitchat.command.CmdForceChat;
import net.katsstuff.old.chitchat.command.CmdMe;
import net.katsstuff.old.chitchat.command.CmdPM;
import net.katsstuff.old.chitchat.command.CmdReply;
import net.katsstuff.old.chitchat.command.CmdShout;
import net.katsstuff.old.chitchat.command.CommandBase;
import net.katsstuff.old.chitchat.helper.LogHelper;
import net.katsstuff.old.chitchat.io.ConfigSettings;
import net.katsstuff.old.chitchat.io.ConfigurateStorage;
import net.katsstuff.old.chitchat.io.IPersistentStorage;
import net.katsstuff.old.chitchat.io.NBTStorage;
import net.katsstuff.old.chitchat.io.StorageType;
import net.katsstuff.old.chitchat.io.StorageTypeSerializer;
import net.katsstuff.old.chitchat.lib.LibPlugin;
import ninja.leaping.configurate.objectmapping.serialize.TypeSerializerCollection;
import ninja.leaping.configurate.objectmapping.serialize.TypeSerializers;

@Plugin(id = LibPlugin.ID, name = LibPlugin.NAME, version = LibPlugin.VERSION, authors = LibPlugin.AUTHORS, description = LibPlugin.DESCRIPTION)
public class ChitChat {

	private static ChitChat plugin;
	private IPersistentStorage storage;
	private ConfigSettings cfg;

	@Inject
	private PluginContainer pluginContainer;

	@Inject
	private Logger log;
	@Inject
	@ConfigDir(sharedRoot = false)
	private Path configDir;

	@Listener
	public void gameConstruct(GameConstructionEvent event) {
		plugin = this;
	}

	@Listener
	public void init(GameInitializationEvent event) {
		DataManager dataManager = Sponge.getDataManager();

		dataManager.register(UserChitChatData.class, ImmutableUserChitChatData.class, new UserChitChatManipulatorBuilder());

		TypeSerializerCollection serializers = TypeSerializers.getDefaultSerializers();
		serializers.registerType(TypeToken.of(ChannelChitChat.ChannelRoot.class), new ChannelRootSerializer());
		serializers.registerType(TypeToken.of(ChannelChitChat.class), new ChannelChitChatSerializer());
		serializers.registerType(TypeToken.of(StorageType.class), new StorageTypeSerializer());

		cfg = new ConfigSettings(configDir);
		storage = createStorage(configDir);

		registerCommand(CmdChannel.INSTANCE);
		registerCommand(CmdShout.INSTANCE);
		registerCommand(CmdPM.INSTANCE);
		registerCommand(CmdReply.INSTANCE);
		registerCommand(CmdMe.INSTANCE);
		registerCommand(CmdChitChat.INSTANCE);
		registerCommand(CmdAnnounce.INSTANCE);
		registerCommand(CmdForceChat.INSTANCE);

		Sponge.getEventManager().registerListeners(this, new ChatListener());
		ChannelChitChat.ChannelRoot.init();
	}

	@Listener
	public void reload(GameReloadEvent event) {
		getConfig().reload();
		storage.loadRootChannel();
	}

	public Logger getLog() {
		return log;
	}

	public static ChitChat getPlugin() {
		return plugin;
	}

	public static IPersistentStorage getStorage() {
		return plugin.storage;
	}

	public static ConfigSettings getConfig() {
		return plugin.cfg;
	}

	public static PluginContainer getPluginContainer() {
		return plugin.pluginContainer;
	}

	/**
	 * Register a command both as a command, and to the help system.
	 */
	private void registerCommand(CommandBase command) {
		Sponge.getCommandManager().register(this, command.getCommand(), command.getAliases());
		command.registerHelp();
	}

	private IPersistentStorage createStorage(Path path) {
		String name = "storage";
		switch(cfg.getStorageType()) {
			case PLAINTEXT:
				return new ConfigurateStorage(path, name);
			case NBT:
				try {
					return new NBTStorage(path, name, cfg.getNbtCompressed());
				}
				catch(IOException e) {
					e.printStackTrace();
					LogHelper.error("Something went wrong when initiating the NBT file for ChitChat. Defaulting to plaintext");
					return new ConfigurateStorage(path, name);
				}
			default:
				LogHelper.error("Something went wrong when getting the storage type to use for ChitChat. Defaulting to plaintext");
				return new ConfigurateStorage(path, name);
		}
	}
}

File: 6.0.0/src/main/scala/net/katsstuff/chitchat/ChitChat.scala
package net.katsstuff.chitchat

import java.nio.file.Path

import org.slf4j.Logger
import org.spongepowered.api.Sponge
import org.spongepowered.api.config.ConfigDir
import org.spongepowered.api.data.{DataQuery, DataRegistration}
import org.spongepowered.api.data.key.{Key, KeyFactory}
import org.spongepowered.api.data.value.mutable.Value
import org.spongepowered.api.effect.sound.{SoundType, SoundTypes}
import org.spongepowered.api.event.Listener
import org.spongepowered.api.event.game.GameReloadEvent
import org.spongepowered.api.event.game.state.{GameConstructionEvent, GameInitializationEvent}
import org.spongepowered.api.plugin.{Dependency, Plugin, PluginContainer}
import org.spongepowered.api.service.permission.Subject
import org.spongepowered.api.text.format.TextColors._

import com.google.inject.Inject

import io.github.katrix.katlib.command.CommandBase
import io.github.katrix.katlib.helper.Implicits._
import io.github.katrix.katlib.lib.LibKatLibPlugin
import io.github.katrix.katlib.serializer.TypeSerializerImpl._
import io.github.katrix.katlib.{ImplKatPlugin, KatLib}
import net.katsstuff.chitchat.chat.{ChannelHandler, ChatListener}
import net.katsstuff.chitchat.chat.channel.{Channel, SimpleChannel}
import net.katsstuff.chitchat.chat.data.{ChannelData, ChannelDataBuilder, ImmutableChannelData}
import net.katsstuff.chitchat.command.{
  CmdAnnounce,
  CmdInChannel,
  CmdJoinChannel,
  CmdListChannel,
  CmdMe,
  CmdModChannel,
  CmdPm,
  CmdReply,
  CmdShout
}
import net.katsstuff.chitchat.lib.LibPlugin
import net.katsstuff.chitchat.persistant.{ChitChatConfig, ChitChatConfigLoader, StorageLoader}
import ninja.leaping.configurate.objectmapping.serialize.{TypeSerializer, TypeSerializers}

object ChitChat {

  final val Version         = s"${KatLib.CompiledAgainst}-2.0.1"
  final val ConstantVersion = "6.0.0-2.0.1"
  assert(Version == ConstantVersion)

  private var _plugin: ChitChat = _
  implicit def plugin: ChitChat = _plugin
}

@Plugin(
  id = LibPlugin.Id,
  name = LibPlugin.Name,
  description = LibPlugin.Description,
  authors = Array("Katrix"),
  version = ChitChat.ConstantVersion,
  dependencies = Array(new Dependency(id = LibKatLibPlugin.Id, version = KatLib.ConstantVersion))
)
class ChitChat @Inject()(logger: Logger, @ConfigDir(sharedRoot = false) cfgDir: Path, spongeContainer: PluginContainer)
    extends ImplKatPlugin(logger, cfgDir, spongeContainer)
    with ChitChatPlugin {

  implicit private val plugin: ChitChatPlugin = this

  private lazy val configLoader            = new ChitChatConfigLoader(configDir)
  private lazy val storageLoader           = new StorageLoader(configDir)
  private lazy implicit val channelHandler = new ChannelHandler(storageLoader)

  private var _config: ChitChatConfig = _
  override def config: ChitChatConfig = _config

  @Listener
  def gameConstruct(event: GameConstructionEvent): Unit = ChitChat._plugin = this

  @Listener
  def init(event: GameInitializationEvent): Unit = {
    _config = configLoader.loadData
    Channel.registerChannelType[SimpleChannel]("simple")
    TypeSerializers.getDefaultSerializers.registerType(typeToken[Channel], implicitly[TypeSerializer[Channel]])

    channelHandler.registry.registerChannelType("simple", {
      case (name, prefix, description, none) if none.trim.isEmpty =>
        Right(new SimpleChannel(name, prefix, description, Set()))
      case _ => Left(t"${RED}This channel does not take extra arguments")
    })

    DataRegistration.builder
      .dataClass[ChannelData, ImmutableChannelData](classOf[ChannelData])
      .immutableClass(classOf[ImmutableChannelData])
      .builder(new ChannelDataBuilder())
      .manipulatorId(s"${LibPlugin.Id}:ChannelData")
      .buildAndRegister(container.container)

    Sponge.getEventManager.registerListeners(this, new ChatListener)

    registerCommand(pluginCmd)
    registerCommand(new CmdModChannel)
    registerCommand(new CmdAnnounce)
    registerCommand(new CmdInChannel)
    registerCommand(new CmdJoinChannel)
    registerCommand(new CmdListChannel)
    registerCommand(new CmdMe)
    registerCommand(new CmdShout)

    val pmCmd = new CmdPm
    registerCommand(pmCmd)
    registerCommand(new CmdReply(pmCmd))
  }

  @Listener
  def reload(event: GameReloadEvent): Unit = {
    configLoader.reload()
    _config = configLoader.loadData
    storageLoader.reload()
    channelHandler.reloadChannels()
  }

  private def registerCommand(cmd: CommandBase): Unit =
    Sponge.getCommandManager.register(this, cmd.commandSpec, cmd.aliases: _*)

  override val versionHelper: VersionHelper = new VersionHelper {
    override val levelUpSound: SoundType = SoundTypes.ENTITY_PLAYER_LEVELUP
    override val ChannelKey: Key[Value[String]] = KeyFactory.makeSingleKey(
      typeToken[String],
      typeToken[Value[String]],
      DataQuery.of(LibPlugin.Id, "channel", "currentChannel"),
      s"${LibPlugin.Id}:currentChannel",
      "ChitChat Current Channel"
    )
    override def getSubjectOption(subject: Subject, option: String): Option[String] = subject.getOption(option).toOption
  }
}
